File: src/main/storage/fuel-file.ts
// fuel.toml — one [[entry]] per fill-up (XTRITIUM §4.4).

import type { FuelEntry } from '../../shared/records.js'
import { formatPump, toPump } from '../../shared/scaled.js'
import {
  emptyDocument,
  readEntryFile,
  writeEntryFile,
  type EntryDocument,
  type EntrySpec
} from './entry-file.js'
import {
  basicString,
  dateLines,
  line,
  readBoolean,
  readDate,
  readInteger,
  readNumber,
  readString
} from './toml.js'

/**
 * `total` is listed here as a key we recognise precisely so that it is DROPPED
 * rather than carried: litres × price is derived and never stored (§4.4). A
 * file that arrives with one written in loses it on the next save.
 */
const KNOWN_KEYS = [
  'id',
  'date',
  'odometer_km',
  'litres',
  'price_per_litre',
  'full_tank',
  'fuel_type',
  'total'
] as const

export const FUEL_SPEC: EntrySpec<FuelEntry> = {
  kind: 'fuel',
  knownKeys: KNOWN_KEYS,
  readEntry: (raw, id) => ({
    id,
    date: readDate(raw['date']),
    odometer_km: readInteger(raw['odometer_km']),
    litres: toPump(readNumber(raw['litres'])),
    price_per_litre: toPump(readNumber(raw['price_per_litre'])),
    full_tank: readBoolean(raw['full_tank']),
    fuel_type: readString(raw['fuel_type'])
  }),
  emitEntry: (entry) => [
    line('id', basicString(entry.id)),
    ...dateLines('date', entry.date),
    line('odometer_km', entry.odometer_km.toString()),
    line('litres', formatPump(entry.litres)),
    line('price_per_litre', formatPump(entry.price_per_litre)),
    line('full_tank', entry.full_tank ? 'true' : 'false'),
    line('fuel_type', basicString(entry.fuel_type))
  ]
}

export type FuelDocument = EntryDocument<FuelEntry>

export function emptyFuel(): FuelDocument {
  return emptyDocument<FuelEntry>()
}

export function readFuel(file: string): FuelDocument {
  return readEntryFile(file, FUEL_SPEC)
}

export function writeFuel(file: string, document: FuelDocument): void {
  writeEntryFile(file, document, FUEL_SPEC)
}

File: android/app/src/main/kotlin/io/github/sudomegas/tritium/storage/FuelFile.kt
package io.github.sudomegas.tritium.storage

import dev.eav.tomlkt.TomlTable
import java.io.File

/** `fuel.toml` — one `[[entry]]` per fill-up (XTRITIUM §4.4), ported from `fuel-file.ts`. */

/**
 * `total` is listed here so it is recognised precisely to be **dropped**,
 * not carried: `litres × price` is derived and never stored (§4.4 — XTRITIUM
 * §3 principle 7). A file arriving with one loses it on the next save.
 */
private val KNOWN_KEYS = setOf(
    "id", "date", "odometer_km", "litres", "price_per_litre", "full_tank", "fuel_type", "total",
)

object FuelSpec : EntrySpec<FuelEntry> {
    override val kind = RecordKind.FUEL
    override val knownKeys = KNOWN_KEYS

    override fun readEntry(table: TomlTable, id: String): FuelEntry = FuelEntry(
        id = id,
        date = readDate(table, "date"),
        odometerKm = readInteger(table, "odometer_km"),
        litres = Scaled.toPump(readNumber(table, "litres")),
        pricePerLitre = Scaled.toPump(readNumber(table, "price_per_litre")),
        fullTank = readBoolean(table, "full_tank"),
        fuelType = readString(table, "fuel_type"),
    )

    override fun emitEntry(entry: FuelEntry): List<String> = buildList {
        add(line("id", basicString(entry.id)))
        addAll(dateLines("date", entry.date))
        add(line("odometer_km", entry.odometerKm.toString()))
        add(line("litres", Scaled.formatPump(entry.litres)))
        add(line("price_per_litre", Scaled.formatPump(entry.pricePerLitre)))
        add(line("full_tank", if (entry.fullTank) "true" else "false"))
        add(line("fuel_type", basicString(entry.fuelType)))
    }
}

typealias FuelDocument = EntryDocument<FuelEntry>

fun emptyFuel(): FuelDocument = emptyDocument()
fun readFuel(file: File): FuelDocument = readEntryFile(file, FuelSpec)
fun writeFuel(file: File, document: FuelDocument) = writeEntryFile(file, document, FuelSpec)
